mmdet3d/core/bbox/coders/km3d_decoder.py

```diff
@@ -0,0 +1,383 @@
+import numpy as np
+import torch
+from torch.nn import functional as F
+
+
+def _nms(heat, kernel=3):
+    """Process heat with Non Maximum Suppression.
+git
+    Args:
+        heat (torch.Tensor): Predictions from model in shape (batch_size *
+            channels * height * width)
+        kernel (int): kernel size for max pooling
+
+    Returns:
+        torch.Tensor: heatmap after NMS, in shape (batch_size * channels *
+        height * width)
+    """
+    pad = (kernel - 1) // 2
+    hmax = F.max_pool2d(heat, (kernel, kernel), stride=1, padding=pad)
+    keep = (hmax == heat).float()
+    return heat * keep
+
+
+def _topk_channel(scores, K=40):
+    """Select top k channels of scores, index, y's, and x's given scores.
+
+    Args:
+        scores (torch.Tensor): batch_size * channels * height * width
+        K (int): top K value
+
+    Returns:
+        torch.Tensor (batch_size * channels * K): top k scores
+        torch.Tensor (batch_size * channels * K): top k index
+        torch.Tensor (batch_size * channels * K): top k y's
+        torch.Tensor (batch_size * channels * K): top k x's
+    """
+    batch, cat, height, width = scores.size()
+
+    topk_scores, topk_inds = torch.topk(scores.view(batch, cat, -1), K)
+
+    topk_inds = topk_inds % (height * width)
+    topk_ys = (topk_inds / width).int().float()
+    topk_xs = (topk_inds % width).int().float()
+
+    return topk_scores, topk_inds, topk_ys, topk_xs
+
+
+def _gather_feat(feat, ind, mask=None):
+    """Gather feat with assigned index, with index expanded to the third
+    dimension.
+
+    Args:
+        feat (torch.Tensor): Predictions from model in shape (batch_size *
+            channels * height * width)
+        ind (torch.Tensor): index to be gathered
+
+    Returns:
+        torch.Tensor: gathered feat, in shape (batch_size * channels *
+        height * width)
+    """
+
+    # expand ind with the third dimension to be the same as feat
+    dim = feat.size(2)
+    ind = ind.unsqueeze(2).expand(ind.size(0), ind.size(1), dim)
+    feat = feat.gather(1, ind)
+    if mask is not None:
+        mask = mask.unsqueeze(2).expand_as(feat)
+        feat = feat[mask]
+        feat = feat.view(-1, dim)
+    return feat
+
+
+def _transpose_and_gather_feat(feat, ind):
+    """gather feat with assigned index, with feat be permuted with (0,2,3,1)
```

Comment:
Rename the function and rewrite the docstring. Although this kind of name is strange, at least renamed to _permute_and_gather_feat